### PR TITLE
8278228: C2: Improve identical back-to-back if elimination

### PR DESCRIPTION
**Pull request by rwestrel**
C2 has had the ability to optimize:

(1)
```
if (some_condition) {
  // body 1
} else {
  // body 2
}
if (some_condition) {
  // body 3
} else {
  // body 4
}
```
into:

(4)
```
if (some_condition) {
  // body 1
  // body 3
} else {
  // body 2
  // body 4
}
```
for a while.

This is achieved by the intermediate step:

(2)
```
if (some_condition) {
  // body 1
  some_condition2 = true;
} else {
  // body 2
  some_condition2 = false;
}
if (some_condition2) {
  // body 3
} else {
  // body 4
}
```
which then allows the use of the exiting split if optimization. As a
result, the graph is transformed to:

(3)
```
if (some_condition) {
  // body 1
  some_condition2 = true;
  if (some_condition2) {
    body3: // a Region here
    // body3
  } else {
    goto body4;
  }
} else {
  // body 2
  some_condition2 = false;
  if (some_condition2) {
    goto body3;
  } else {
    body4:  // another Region here
    // body4;
  }
}
```
and finally to (4) above.

Recently, 8275610 has shown that this can break if some_condition is a
null check. If, say, body 3 has a control dependent CastPP, then when
body 1 and body 3 are merged, the CastPP of body 3 doesn't become
control dependent on the dominating if (because in step (2), the
CastPP hides behind a Region). As a result, the CastPP loses its
dependency on the null check.

After discussing this with Christian, it seemed this was caused by the
way this transformation relies on split if: having custom code that
wouldn't create Regions at body3 and body4 that are then optimized out
would solve the problem. Anyway, after looking at the split if code,
trying to figure out how to tease it apart in smaller steps and
reusing some of them to build a new transformation, it seemed too
complicated. So instead, I propose reusing split if in a slightly
different way:

skip step (2) but perform split if anyway to obtain:
```
if (some_condition) {
  // body 1
  if (some_condition) {
    body3: // Region1
    // CastPP is here, control dependent on Region1
    // body3
  } else {
    goto body4;
  }
} else {
  // body 2
 if (some_condition) {
    goto body3;
  } else {
    body4:  // Region2
    // body4;
  }
}
```
- A CastPP node would still be behind a Region. So next step, is to push
control dependent nodes through Region1 and Region2:
```
if (some_condition) {
  // body 1
  if (some_condition) {
    // A CastPP here
    body3: // Region1
    // body3
  } else {
    goto body4;
  }
} else {
  // body 2
 if (some_condition) {
    // A CastPP here
    goto body3;
  } else {
    body4:  // Region2
    // body4;
  }
}
```
- And then call dominated_by() to optimize the dominated:

if (some_condition) {

in both branches of the dominating if (some_condition) {. That also
causes the CastPP to become dependent on the dominating if.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8278228](https://bugs.openjdk.java.net/browse/JDK-8278228): C2: Improve identical back-to-back if elimination

### Reviewers
 * [Christian Hagedorn](https://openjdk.java.net/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6882/head:pull/6882` \
`$ git checkout pull/6882`

Update a local copy of the PR: \
`$ git checkout pull/6882` \
`$ git pull https://git.openjdk.java.net/jdk pull/6882/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6882`

View PR using the GUI difftool: \
`$ git pr show -t 6882`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6882.diff">https://git.openjdk.java.net/jdk/pull/6882.diff</a>

</details>

### CODE REVIEW

**Comment by rwestrel** [comment]
I pushed a tweak to PhaseIdealLoop::push_pinned_nodes_thru_region() so it's applied only to nodes that will actually be rewired by dominated_by() and not all pinned nodes.

**Review by chhagedorn — APPROVED** [review]
That's a nice solution to use split if like that! Looks good.

Maybe you want to add the example (1) and the final result in the last snippet as comment to the optimization as well to provide a visual hint.

**Comment by rwestrel** [comment]
> That's a nice solution to use split if like that! Looks good.
> 
> Maybe you want to add the example (1) and the final result in the last snippet as comment to the optimization as well to provide a visual hint.

Thanks. I should have addressed your comments and suggestions with the new version. I also moved all that logic in its own method.

File: test/hotspot/jtreg/compiler/c2/irTests/TestBackToBackIfs.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (c) 2021, Red Hat, Inc. All rights reserved.
```
Comment (vnkozlov) [inline test/hotspot/jtreg/compiler/c2/irTests/TestBackToBackIfs.java:2]: Year 2022.

**Review by vnkozlov — COMMENTED** [review]
Looks good to me.

@chhagedorn please run testing.

**Review by chhagedorn — COMMENTED** [review]
Thanks for coming back with the updates! That's even better to move it to a separate method.

Two more things I've noticed:
https://github.com/openjdk/jdk/blob/bc12381105ef5ba14f99304a220817e97c9a99b5/src/hotspot/share/opto/split_if.cpp#L445
This could also be changed to `RegionNode*` with an assertion moved from
https://github.com/openjdk/jdk/blob/bc12381105ef5ba14f99304a220817e97c9a99b5/src/hotspot/share/opto/split_if.cpp#L37

And `new_iff` could also be changed to `RegionNode*`:
https://github.com/openjdk/jdk/blob/bc12381105ef5ba14f99304a220817e97c9a99b5/src/hotspot/share/opto/split_if.cpp#L494

Will submit some testing and get back with the results.

**Comment by rwestrel** [comment]
> Two more things I've noticed:

Thanks Christian. Does the last commit cover your comments?

File: src/hotspot/share/opto/split_if.cpp [hunk 2]
```diff
@@ -31,7 +31,7 @@
 
 //------------------------------split_thru_region------------------------------
 // Split Node 'n' through merge point.
-RegionNode* PhaseIdealLoop::split_thru_region(Node* n, Node* region) {
+RegionNode* PhaseIdealLoop::split_thru_region(Node* n, RegionNode* region) {
   assert(n->is_CFG(), "");
   assert(region->is_Region(), "");
```
Comment (chhagedorn) [inline src/hotspot/share/opto/split_if.cpp:36]: This assertion can also be removed or moved to L445 as sanity check.

**Review by chhagedorn — APPROVED** [review]
That looks good, thanks for doing all the updates! I think the code is now easier to read.

Testing on Linux and Windows looked good. Mac is still running.

File: src/hotspot/share/opto/split_if.cpp (hunk 2 above)
Comment (rwestrel) [reply src/hotspot/share/opto/split_if.cpp]: Right, as_Region() takes care of the assert anyway.

**Review by chhagedorn — APPROVED** [review]

**Review by vnkozlov — APPROVED** [review]
Looks good.

**Comment by chhagedorn** [comment]
Testing completed and looked good!

**Comment by rwestrel** [comment]
@chhagedorn @vnkozlov thanks for the reviews and testing.

**Pull request closed without merging**